### ISSUE TITLE
Keep update status while rebooting

### ISSUE THREAD

**Issue by jjcarstens** [nerves-hub/nerves_hub_web#494]
A good point @fhunleth brought up: When a device gets a fw update, it goes through the whole process and you get a nice status bar. But on reboot, the device just shows as "offline" while its coming back up with the new fw.

It would be nice to keep a status around to signify the device is offline perform update vs just normal offline. That would help in quick checks to tell if a device is offline because something failed in the update or if it has _always_ been offline without ever getting the update.

Note, you can probably figure this out by clicking in and looking at the audits for the device, so this is definitely a small enhancement and "nice to have"

**Comment by danielspofford** [comment]
Perhaps this is best facilitated by a `NervesHub.reboot[_device]/0` or something. Basically a generic way for the device to be able to reboot itself that would also inform NervesHubWeb that the device is about to go offline but has the intention of coming back online. It could be leveraged post-fw-download as well as whenever you are rebooting for any reason.

**Comment by joshk** [comment]
I started looking into this as it is a good feature to have, although it isn't clean-cut.

Based on our current code, the Device sends a message to Hub to say its rebooting and Hub listens for this but does nothing. 

We don't store 'connected' or 'disconnected' anywhere in particular. Instead, we poll the running instances to see if the Device is connected.

And we can't store this information in the Registry, as the Device information in the Registry is purged when the Device disconnects, which will happen when it reboots.

Plus, if we were to store 'rebooting' somewhere, we likely want that to expire after x minutes if the Device doesn't come back online. I think it would also be sane to add an Audit Log message that the Device reboot didn't complete successfully.

To summerize:
- we don't store the connection state in the DB or ETS
- we use PubSub broadcast to check if the Device is currently connected (`DeviceChannel` replies)
- if we were to store information related to the Device rebooting in Registry, where we store other transient Device information, it would be purged when the Device disconnects during the reboot
- where ever we do store it, we need to expire that information after X period of minutes and turn that 'expected reboot' into a 'reboot failed' (we didn't see the device come back online)

**Comment by oestrich** [comment]
Some extra context for this:

- We used to track via a global registry of sorts but some instances of nerves hub got too big for it to work
- We don't store online status in the database because it's easy to go stale, that's why we have the live check which isn't great but it's guaranteed to be online if it says it is
- The one registry in use for devices is related to deployments for the orchestrator and is expected to be local because each node has an orchestrator per deployment loaded

**Comment by lawik** [comment]
@elinol does the new device connection stuff capture enough information to do this?

**Comment by elinol** [comment]
It's possible to pass a reason along with the `:disconnected` status. I don't know if a third status option, like `:rebooting` would be helpful?

**Comment by nshoes** [comment]
@elinol do we set a `reason` for any other status? If not, I'd vote for `:rebooting`.

**Comment by elinol** [comment]
@nshoes I'm not too involved in how disconnecting reasons are passed, but I can find a matching on `{:shutdown, :disconnected}` in `device_socket.ex`, and when reconnecting my little local test device I get reason `{:error, :closed}`

**Comment by nshoes** [comment]
@elinol Right! I did some digging, we'll need to [update nerves_hub_link](https://github.com/nerves-hub/nerves_hub_link/pull/249), [match here](https://github.com/nerves-hub/nerves_hub_web/blob/5fc990c5205aaaffddf70edad1f7253ac69176a9/lib/nerves_hub_web/channels/device_channel.ex#L363) and then we can flag the device properly.

**Comment by jjcarstens** [comment]
I believe we should be able to capture this based on previous state. I.e if we are in an updating status, the next offline message should assume "rebooting"

It doesn't hurt to send the message from device, but we might be racing message delivery and socket close so prob best to find the best effort assertion on web

**Comment by lawik** [comment]
@joshk this is the case now right? I know you talked about having way more status information at least.

**Comment by joshk** [comment]
This has been addressed with some recent work on richer in-flight update statuses.
